### PR TITLE
Added symlink check for zip decompression

### PR DESCRIPTION
**Pull request by bytecook**
Supports #242
(hopefully)

### CODE REVIEW

**Comment by coolaj86** [comment]
I'm going to merge in a few other commits and then ask you to do a `git rebase origin master`.

There may be one conflict in terms of updating an error handler, but I don't think much else.

**Comment by bytecook** [comment]
@coolaj86 I was supposed to wait for your go ahead to run `git rebase origin master`, correct?

**Comment by coolaj86** [comment]
@bytecook All set and ready for you to rebase.

- `git pull --rebase https://github.com/mholt/archiver master` to replay your commits after the current master
- `git rebase -i` to squash away excess commits
- `git push --force` to force the new, rewritten history of this PR

Would you mind making a commit for https://github.com/mholt/archiver/pull/246#issuecomment-703280843 as well? It's just a two line change.

**Comment by bytecook** [comment]
Ok, I'll get on it.

**Comment by coolaj86** [comment]
Hmmm... that did not go well. Try this:

```bash
git clone https://github.com/bytecook/archiver.git
pushd archiver/
git remote add upstream https://github.com/mholt/archiver.git

git reset --hard 097427f769ad76fc0ed3d9d3a28aed9406451237
git checkout -b fixup
git reset --hard HEAD~20
git pull upstream master

git checkout master
git rebase -i fixup
# edit the commit to "squash" some of the redundant commits

git push --force origin master
```

**Comment by coolaj86** [comment]
Sidenote: if you're a vim user, I'd recommend using vim-go:

```bash
curl https://webinstall.dev/vim-go | bash
```

Though perhaps the formatting issues are just a matter of bad space changes through the merge/rebase.

**Comment by bytecook** [comment]
let me try again

**Comment by coolaj86** [comment]
@bytecook I was looking at this more closely, and this actually does not check symlinks at all.

Take a look at [os.Readlink](https://golang.org/pkg/os/#Readlink) and [os.FileInfo](https://golang.org/pkg/os/#FileInfo).

**Comment by coolaj86** [comment]
Also, I'll make a little video showing how to do the rebase and squash. It seems that you've tried to mix several branches together in your master, and the commit history is very messy. I'll show you how to fix it.

**Comment by coolaj86** [comment]
How to Rebase and Squash:

["Git Rebase & Squash"](https://i.imgur.com/BBu8iSL.jpg)

File: filecompressor.go [hunk 1]
```diff
@@ -52,14 +52,11 @@ func (fc FileCompressor) DecompressFile(source, destination string) error {
...
-	//fd.Name() is only a valid function call if there is a path to fd.
-	//Lstat only returns os.PathErrors.
-	//ok checks if there is a path before checking if symlink is attached to destination.
-	if _, ok := err.(*os.PathError); !ok {
-		if fd.Name() == destination {
-			return fmt.Errorf("destination is symlink: %s", destination)
-		}
+	// os.Readlink returns the destination of the named symbolic link. 
+	// If there is an error, it will be of type *PathError.
+	sym, err := os.Readlink(destination)
```
Comment (coolaj86) [inline filecompressor.go:57]: This is not correct.
Readlink returns either a string that is the name of the link's target, or an error if the file is not a link.

Let's take a step back. Why don't you first communicate what you think the code should do via comments on the issue rather than trying different things in code. Do you understand what symlinks are and how they work?

**Review by coolaj86 — CHANGES_REQUESTED** [review]
